Created AppManager class

albertoursino opened the pull request:
This class is useful for managing all those previously merged.

It allows you to **send RingCommands and manage those received**.
Through this class you can also, thanks to the "checkPassword" method, check if the password contained in the RingCommand corresponds to the one saved in memory.

Commit:
Created the AppManager class

Ready to send in pull request

Commit:
File remmoved

Useless file.
I forgot to take it off.

CremaLuca commented on app/src/main/java/com/eis/phoneringer/AppManager.java:
```diff
@@ -0,0 +1,96 @@
+     * @param ringCommand a ring command not null
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    public void onRingCommandReceived(Context context, RingCommand ringCommand, final Ringtone ringtone) {
```
If you really **really** want the `ringCommand` to be not null you could add the @NonNull annotation. It's up to you, not necessary (you could also add it on context)

Commit:
File Removed

Useless file.
I forgot to take it off.

CremaLuca commented on app/src/main/java/com/eis/phoneringer/AppManager.java:
```diff
@@ -0,0 +1,96 @@
+     * If the password of the message received is valid then play ringtone for fixed amount of time
+     *
+     * @param context     of the application
+     * @param ringCommand a ring command not null
```
Missing `ringtone` @param description

CremaLuca commented on app/src/main/java/com/eis/phoneringer/AppManager.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    /**
+     * Method used to send a RingCommand via SMS using the library class {@link SMSHandler}
```
Love the {@link}

albertoursino replied:
Who knows from whom I learned it.........

CremaLuca commented on app/src/main/java/com/eis/phoneringer/AppManager.java:
```diff
@@ -0,0 +1,96 @@
+     * @param ringCommand a valid RingCommand object
+     * @return a boolean: true = passwords are corresponding, false = passwords are NOT corresponding
+     */
+    public boolean checkPassword(Context context, RingCommand ringCommand) {
```
Is this method **_really_** `public`? Can't this become `private`?

albertoursino replied:
You may be right.
I am not very convinced because it is a particular method that could also be used in other classes, but for now I put it as private as you suggest because it is actually used only within AppManager.

CremaLuca commented on app/src/main/java/com/eis/phoneringer/AppManager.java:
```diff
@@ -0,0 +1,96 @@
+     * @param context     of the application
+     * @param ringCommand to send
+     * @throws InvalidTelephoneNumberException
+     * @throws InvalidSMSMessageException
```
Are these two exceptions actually thrown? `parseCommand` doesn't throw anything, even though it probably should because we don't check `SMSPeer` validity in RingCommand

CremaLuca replied:
You should probably go and edit RingCommandHandler.parseCommand and add `throws InvalidTelephoneNumberException, InvalidSMSMessageException`

albertoursino replied:
These exceptions can only be launched in this method because the SMSHandler is invoked only within it.

CremaLuca replied:
`SMSHandler` doesn't throw any exception. Those who throw those two exceptions are `SMSPeer` and `SMSMessage`, which are created inside `RingCommandHandler.parseCommand`, so that method should throw them.

albertoursino replied:
Ops, you're right.
Is the _ringCommandhandler_ class, when creating the SMSMessage, which could launch these exceptions.

albertoursino replied:
Well we have to update the RingCommandHandler class.

Commit:
Added some @NonNull annotations and fixed some specifications

CremaLuca reviewed (changes requested):
A few changes, mostly for code reliability

Commit:
Resolved the review by Luca Crema

Commit:
Removed exceptions

CremaLuca commented on app/src/main/java/com/eis/phoneringer/structure/AppManager.java:
```diff
@@ -0,0 +1,96 @@
+     * @param context     of the application
+     * @param ringCommand to send
+     */
+    public void sendCommand(Context context, @NonNull RingCommand ringCommand) throws InvalidSMSMessageException, InvalidTelephoneNumberException {
```
Nah man, the two exceptions are still being thrown because this method calls `RingCommandHandler.parseCommand()` which throws them.
Also, you removed the specifications but not the "throws" after the method signature. Bring the specifications back and give them a description of when they are thrown

AlbertoUrsino replied:
I know, corrected :D

albertoursino replied:
Ok I removed the "throws" but wait, why should I add the exceptions back and specify that they are launched by the method of **another class**?

CremaLuca replied:
Yes, you should specify that those exceptions are thrown because you make a call to `RingCommandHandler.parseCommand()` and that may throw exceptions

albertoursino replied:
Well ok, I put the exceptions back.
Do we leave the RingCommandHandler class as it is?

CremaLuca replied:
For now yes, you can't edit it from this branch

Commit:
Small fix to a specification

Commit:
Rewrote exception in the "sendCommand" method

Cogno-Marco reviewed (approved):
Approving changes, quick little class not doing much 👍

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/structure/AppManager.java:
```diff
@@ -0,0 +1,98 @@
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresApi;
+
+import com.eis.smslibrary.SMSHandler;
+import com.eis.smslibrary.exceptions.InvalidSMSMessageException;
+import com.eis.smslibrary.exceptions.InvalidTelephoneNumberException;
```
You're importing classes from package `com.eis.smslibrary` (which is located in another module), but you didn't include its module as a dependency in `app`'s `build.gradle`.

giovannivelludo commented on app/src/main/java/com/eis/phoneringer/structure/AppManager.java:
```diff
@@ -0,0 +1,98 @@
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    public void onRingCommandReceived(Context context, @NonNull RingCommand ringCommand, final Ringtone ringtone) {
+        final RingtoneHandler ringtoneHandler = RingtoneHandler.getInstance();
```
Class RingtoneHandler does not exist in the [branch](https://github.com/AlbertoUrsino/killer-app/tree/03d65817ed7d2879e3c0e9595db7794fd9f0f3e0) you're requesting a pull from, how could you check if your code even compiles?

albertoursino replied:
I simply add all the necessary files from the outside and then only commit the class that has to go to pull request.
All these "outsiders" files will be stashed in my local repo.

albertoursino replied:
Of course the files I use are:

- https://github.com/Cogno-IDU/killer-app
- https://github.com/Cogno-IDU/smslibrary

giovannivelludo replied:
Why don't you just pull from Cogno-IDU/killer-app/master to AlbertoUrsino/killer-app/AppManager before making any changes, so that you get a functioning app in your branch?
I don't exactly understand why you used git stash to put aside the "outsider" files needed to have a working app: from what I understand stashing is used when you're working on a branch and you have to switch to another, but you don't want to lose the changes you made to your working directory and you also don't want to commit them.

albertoursino replied:
If I merge all the "outsider" files in my master and then call a pull request.. Well, I tried but in the pull request will being show all the "ousiders" files as "changed files".. I don't know how to resolve this problem, I tried with "cherry-commits" but I didn't understand them so I opted for this method and it works.

giovannivelludo replied:
> Well, I tried but in the pull request will being show all the "ousiders" files as "changed files"

Could you elaborate on this? From what branch and into what branch did you pull (let's use the term "origin" to refer to your GitHub fork of killer-app, and "upstream" to refer to Cogno-IDU's killer-app)?
Maybe upload a video on YouTube showing what happens in Android Studio when you merge upstream/master into your local branch AppManager, then push to origin/AppManager, then make a pull request from AlbertoUrsino:AppManager to Cogno-IDU:master?

Cogno-Marco replied:
What should happen is that you update the submodule from the commit it's on to a new commit based on the central repo. Since you're updating these you'll see the smslibrary submodule files as changed (since they in fact are changed). This is not a problem if they get in the pull request, you just have to update us that you updated the submodule, so we know that any file related to smslibrary is already ok since we reviewed it in it's own repository

giovannivelludo reviewed (changes requested):
I don't get how you were able to add this class and check that everything works fine, when there's code that references classes that don't exist on your branch.

Commit:
Updated the project with the central repo

Commit:
Merge remote-tracking branch 'refs/remotes/centralRepo/master' into AppManager

AlbertoUrsino commented:
If you notice in the last commit: apart from being the AppManager class we are discussing, there is also the RingtoneHandler class that I moved to the right "structure" package.
If this merge is successful then the "ringCommands" package should be deleted from the central repo.

Commit:
Synchronized the grandle file and updated the submodule

- Also fixed an import on the PasswordManager class

Cogno-IDU commented:
Ready to merge?

albertoursino commented on app/src/main/java/com/eis/phoneringer/structure/AppManager.java:
```diff
@@ -0,0 +1,98 @@
+                }
+            }, TIMEOUT_TIME);
+        } else {
+            Toast.makeText(context, WRONG_PASSWORD_TOAST, Toast.LENGTH_SHORT).show();
```
This toast does not allow us to test the class on "onRingCommandReceived" method and in addition it is not nice to have a toast inside a class and not in the mainactivity.

What do u think?

albertoursino replied:
![Untitled](https://user-images.githubusercontent.com/56396708/70058864-00e3ad00-15e0-11ea-8860-43f6dc70d5ac.png)

What about this solution? Now we can do tests and also there's a better correctness of the code...

Cogno-Marco replied:
Beautiful solution changing from a toast to an exception. I would make just a little edit, from `WrongPasswordException` to `InvalidPasswordException`. Same meaning, different way to say it, which in my opinion is more inline with the standard java api (think about `IllegalArgumentException`)

tonin-alessandra replied:
I think this is a better solution than toast, but I also agree with @Cogno-IDU  about the exception name.

CremaLuca replied:
I don't think `IllegalPasswordException` is the name we're looking for, this code checks if the password is the same, not if it's valid / with the right format. Anyway the name is still kinda intuitive and both can be used, but I'd prefer `Wrong`

ps. I don't like at all the use of the adversative "but" in @tonin-alessandra comment: you are linking two positive sentences with a "but"! That's not correct! Should have used an "and"

albertoursino replied:
After receiving another revision, perhaps from @giovannivelludo, I will fix these expressions.

Cogno-Marco replied:
@CremaLuca I also chose `Illegal` as a keyword because of the specifications
> Exception thrown when the password received is not valid

Not valid, as in, Invalid

In the end it comes down to personal preferences but I'll just make it consistent

CremaLuca replied:
Agreed on `Ìllegal`

albertoursino replied:
Well, now you are in 3 that you want `Illigal` instead of `Wrong` so i'll fix it now.

Commit:
Created "WrongPasswordException" class and tests

- Now the AppManager method "onRingCommandReceived" throws an exception if the password is wrong
- Now the AppManager method "sendCommand" wants a "SMSSentListener" in addiction to the params
- Created tests for the class AppManager

Removed some classes:
- Example tests
- The package "RingCommans" no more exists -> all the classes are now in the "structure" package

Synched and compiled, all work fine!

albertoursino reviewed (commented):
I've done this modifications in the commit  aae0877

Commit:
Updated smslibrary

Commit:
Merge remote-tracking branch 'refs/remotes/CentralRepo/master' into AppManager

With this merge i have resolved conflicts with the preferences module which is now implemented through jitPack

AlbertoUrsino commented:
I did this 9fc23df commit because I tried on another PC to check this branch and check if everything was fine, but there were some errors caused by "smslibrary" which was not updated.

Now it should be all 100% working, I also merged the central repo with this commit c367ca0 in order to resolve some conflicts.

Try this branch in your computer and control if it works fine.

CremaLuca commented on app/src/androidTest/java/com/eis/phoneringer/structure/AppManagerTest.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @Test
+    public void onRingCommandReceived_wrong_password() {
```
What about using
`@Test(expected = WrongPasswordException.class)`
and remove the try-catch? The current method of doing assert.fail to assert that exceptions are thrown is good only if exceptions are NOT Runtime, so when you **must** catch them. In this case, where the exception is runtime, we can use the expected.

CremaLuca commented on app/src/androidTest/java/com/eis/phoneringer/structure/AppManagerTest.java:
```diff
@@ -0,0 +1,53 @@
+            appManager.getInstance().onRingCommandReceived(context, new RingCommand(new SMSPeer(VALID_PHONE_NUMBER), VALID_PASSWORD), ringtoneHandler.getDefaultRingtone(context));
+            //Success
+        } catch (WrongPasswordException e) {
+            Assert.fail("Should hasn't thrown the WrongPasswordException");
```
Test automatically fails if an exception is thrown, no need for this catch

CremaLuca commented on app/src/main/java/com/eis/phoneringer/structure/AppManager.java:
```diff
@@ -0,0 +1,100 @@
+     * @param ringtone    A valid ringtone to be played
+     * @throws WrongPasswordException Exception thrown when the password received is not valid
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
```
1. Why is this required? 
2. What happens if someone has KitKat? Not saying that you should handle that case, but probably if this method doesn't work and there is no alternative method for lower APIs, the whole app doesn't make sense either. It'd probably be good to say that minimum API required is 5.0 on `build.gradle` and remove this annotation

albertoursino replied:
Method `playRingtone` of the class `RingtoneHandler` requires that API

albertoursino replied:
I'm fixing that problem with a modification in the file `build.gradle` as you said

albertoursino replied:
API 5.0 doesn't exist. Minimun level is 7...

CremaLuca replied:
Man, 5.0/Lollipop is build no. 21 https://source.android.com/setup/start/build-numbers

albertoursino replied:
lvl 21 maybe

albertoursino replied:
Sorry I was thinking about the levels not the Api version

CremaLuca reviewed (changes requested):
Nice, mostly good except for a little clarification on handling exceptions in test methods and a weird RequiresApi that I don't understand which line requires it.

Commit:
Fixed some things based on the Crema's review

CremaLuca reviewed (approved):
Now it's perfect

Commit:
Changed "WrongPasswordException" in "IllegalPasswordException"

Cogno-Marco commented on app/src/main/java/com/eis/phoneringer/structure/AppManager.java:
```diff
@@ -0,0 +1,99 @@
+            }, TIMEOUT_TIME);
+        } else {
+            throw new WrongPasswordException();
+        }
```
A quick suggestion (even though it's not necessary, you might still find this useful)
There's a way of programming where you focus on having the parameters you want.
Basically code like this
```
if(condition){
    //code
}
else {
    return;
}
```
Becomes
```
if( !condition){
    return;
}
//code
```

This way long nested if chains just become linear
```
if( c1){
    if( c2){
        if( c3){
            //code
        }
    }
}
```
becomes
```
if( !c1)
   return;
if( !c2)
    return;
if( !c3)
    return;

//code
```

In your case it would be
```
if ( !checkPassword(context, ringCommand)) throw new WrongPasswordException();

//code inside the if
```

Just a good way to write code in my opinion.

albertoursino replied:
Do you mean something like that?

![Untitled](https://user-images.githubusercontent.com/56396708/70151069-5b960b00-16ab-11ea-8d02-2ef62e9d7010.png)

albertoursino replied:
I can change if you prefer.
In my opinion you are in right @Cogno-IDU, this _code writing method_ significantly speeds up both the method and, in the long run, the application itself.


Also makes the code even nicer...

Cogno-Marco replied:
Nah you don't need to change it. I just thought you didn't know about it, or maybe you didn't pay attention, so I just gave some feedback 😄 .
Also yeah, it's a pretty useful coding style, it even has it's own name. Something like... defensive programming?

Cogno-Marco reviewed (approved):
Nothing to say, good work 👍

Commit:
Improved code writing as suggested by Cognolato

I also wrote some comments to make the code more readable

AlbertoUrsino commented:
Well, now I think you can merge it
